upf/walkers/identitydag.py

```diff
@@ -0,0 +1,78 @@
+# Copyright 2021 AIPlan4EU project
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+import upf.environment
+import upf.walkers as walkers
+import upf.operators as op
+from upf.fnode import FNode
+
+class IdentityDagWalker(walkers.DagWalker):
+    """This class traverses a formula and rebuilds it recursively
+    identically.
+    This could be useful when only some nodes needs to be rewritten
+    but the structure of the formula has to be kept.
+    """
+
+    def __init__(self, env: 'upf.environment.Environment', invalidate_memoization=False):
+        walkers.DagWalker.__init__(self, invalidate_memoization)
+        self.env = env
+        self.manager = env.expression_manager
+
+
+    def walk_and(self, args, **kwargs):
```

Comment:
The following methods have the following sign:
```suggestion
    def walk_and(self, expression, args, **kwargs):
```